src/trigger/BufferManager.hpp

```diff
@@ -0,0 +1,164 @@
+/**
+ * @file BufferManager.hpp
+ *
+ * This is part of the DUNE DAQ Application Framework, copyright 2021.
+ * Licensing/copyright details are in the COPYING file that you should have
+ * received with this code.
+ */
+
+#ifndef TRIGGER_SRC_TRIGGER_BUFFERMANAGER_HPP_
+#define TRIGGER_SRC_TRIGGER_BUFFERMANAGER_HPP_
+
+#include "dataformats/Types.hpp"
+
+#include <atomic>
+#include <set>
+
+namespace dunedaq {
+namespace trigger {
+
+/**
+ * @brief BufferManager description.
+ *
+ */
+template<typename BSET>
+class BufferManager
+{
+public:
+  BufferManager();
+  BufferManager(long unsigned int buffer_size)
+    : m_buffer_max_size(buffer_size)
+    , m_buffer_earliest_start_time(0)
+    , m_buffer_latest_end_time(0)
+  {}
+
+  virtual ~BufferManager() {}
+
+  void set_buffer_size(long unsigned int size) { m_buffer_max_size = size; }
+  void clear_buffer() { m_txset_buffer.clear(); }
+  long unsigned int get_buffer_size() { return m_buffer_max_size; }
+  long unsigned int get_stored_size() { return m_txset_buffer.size(); }
+
+  BufferManager(BufferManager const&) = delete;
+  BufferManager(BufferManager&&) = default;
+  BufferManager& operator=(BufferManager const&) = delete;
+  BufferManager& operator=(BufferManager&&) = default;
+
+  /**
+   *  add a TxSet to the buffer. Remove oldest TxSets from buffer if we are at maximum size
+   */
+  bool add(BSET& txs)
+  {
+    if (m_txset_buffer.size() >=
+        m_buffer_max_size) // delete oldest TxSet if buffer full (and updating earliest start time) -> circular buffer
+    {
+      auto firstIt = m_txset_buffer.begin();
+      m_txset_buffer.erase(firstIt);
+      firstIt++;
+      m_buffer_earliest_start_time = (*firstIt).start_time;
```

Comment:
m_buffer_earliest_start_time = m_txset_buffer.begin().start_time.begin(); 

After the erase() instead of iterator++? One line less, don't know which is quicker.

Reply:
Not sure which is quicker, but your suggestion make the code cleaner. Thanks! Fixed (ed511cecccd53868dabb87a8fcad02859c0f7a5a).